parser.go

```
@@ -0,0 +1,191 @@
+
+// LoadFromFile loads ini file
+// Saves all lines locally into an array of strings
+func (parsedMap *Parser) LoadFromFile(fileName string) error {
+	var iniLines []string
+	var input io.Reader
+
+	file, err := os.Open(fileName)
+	if err != nil {
+		ErrCouldNotOpen = fmt.Errorf("cannot open file: %s", fileName)
+		return ErrCouldNotOpen
+	}
+	defer file.Close()
+
+	input = file
+	scanner := bufio.NewScanner(input)
+	for scanner.Scan() {
+		iniLines = append(iniLines, scanner.Text())
+	}
+	return parsedMap.parserLogic(iniLines)
+}
+
+// LoadFromString loads ini script from a string
+// Saves all lines locally into an array of strings
+func (parsedMap *Parser) LoadFromString(str string) error {
+	iniLines := strings.Split(str, "\n")
+	return parsedMap.parserLogic(iniLines)
+}
+
+func (parsedMap *Parser) parserLogic(iniLines []string) error {
+	var section, key, value string
+
+	for _, line := range iniLines {
+		if len(line) == 0 {
+			continue
+		} else if line[0] == '[' {
+			closingParaFound := false
+			for j, ch := range line {
+				if ch == ']' {
+					section = line[1:j]
+					section = strings.Trim(section, " ")
+					closingParaFound = true
+					if parsedMap.dictionary[section] == nil {
+						parsedMap.dictionary[section] = make(map[string]string)
+					}
+				}
+			}
+			if !closingParaFound {
+				ErrSectionNameMissingClosure = fmt.Errorf("section [%s] is missing closure paranthesis ']'", line[1:])
+				return ErrSectionNameMissingClosure
+			}
+			if parsedMap.dictionary == nil {
+				parsedMap.dictionary = make(map[string]map[string]string)
+			}
+
+		} else if line[0] == ']' {
+			ErrWrongParanthesisOrder = fmt.Errorf("wrong section paranthesis order, the section provided, %s, cannot start by ]", line)
+			return ErrWrongParanthesisOrder
+		} else if line[0] == ';' || line[0] == '#' || line[0] == ' ' || line[0] == '\n' || line[0] == '\t' {
+			continue
+		} else {
+			equalFound := false
+			for j, ch := range line {
```

you don't need to loop over all the chars of a line

---

> you don't need to loop over all the chars of a line

I tried thinking of simpler ways but there's no other way to find it
assuming it would be at the end is a special case only

---

you can split with "="

---

> you can split with "="

if there wasn't an = , I wouldn't know then, that's the problem. w that's an error we have to report
as for complexity i think the split function does also an interaction over characters

---

if there wasn't '=' then the resulting slice would be of length 1, you can report an error then

---

Done